### PR TITLE
feature #862: Added the ability to get all users

### LINKED ISSUE #69: Add the ability to get all users
You need to implement an endpoint to get all students. Requirements:
- **search** (by username and email)
- **sorting** (by username and email)
- **filter** (by state - multichoice)
- **pagination**

**Response** - id, username, email, state

image

### LINKED ISSUE #862: Add the ability to get all users
You need to implement an endpoint to get all students. Requirements:
- **search** (by username and email)
- **sorting** (by username and email)
- **filter** (by state - multichoice)
- **pagination**

**Response** - id, username, email, state

image

### PR DESCRIPTION
**Pull request by dmykyr**
- added users.get permission
- added getAll method in UserController
- added getAll method in UserMapper
- added getAll method in UserService
- added count method in UserRepository
- updated findMany method in userRepository
- created QueryAllUsersDTO and SortQAUParam
- created UsersResponse
- updated validationOptionMsg in GLOBALS (by @f-ogoog)

closes fictadvisor/fictadvisor#69

### CODE REVIEW

File: src/v2/api/controllers/UserController.ts [hunk 1]
```diff
@@ -392,6 +385,35 @@ export class UserController {
...
+    description: `\n
+    InvalidQueryException:
+      Page must be a number
+      PageSize must be a number
+      Wrong value for order
+      Sort must be an enum
+      State must be an enum`,
+  })
+  @ApiEndpoint({
+    summary: 'Get information about each user',
```
Comment (f-ogoog) [inline src/v2/api/controllers/UserController.ts]: i think it will be better write: Get all user for admin

File: src/v2/api/dtos/SortQAUParam.ts [hunk 2]
```diff
@@ -0,0 +1,4 @@
+export enum SortQAUParam {
+    USERNAME  = 'username ',
+    EMAIL = 'email',
+}
```
Comment (f-ogoog) [inline src/v2/api/dtos/SortQAUParam.ts]: transfer this enum to QueryAllUsersDTO

File: src/v2/api/dtos/QueryAllUsersDTO.ts [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+import { State } from '@prisma/client';
+import { QueryAllDTO } from '../../utils/QueryAllDTO';
+import { validationOptionsMsg } from '../../utils/GLOBALS';
+import { SortQARParam } from './SortQARParam';
+import { SortQAUParam } from './SortQAUParam';
+import { Transform } from 'class-transformer';
+
+export class QueryAllUsersDTO extends QueryAllDTO {
+    @ApiPropertyOptional({
+      enum: SortQARParam,
```
Comment (f-ogoog) [inline src/v2/api/dtos/QueryAllUsersDTO.ts]: should be SortQAUParam

File: src/v2/api/dtos/QueryAllUsersDTO.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+      description: 'Sorting by field',
+      default: 'displayName',
```
Comment (f-ogoog) [inline src/v2/api/dtos/QueryAllUsersDTO.ts]: SortQAUParam doesn't have this field

File: src/v2/api/dtos/QueryAllUsersDTO.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    })
+    @IsEnum(SortQAUParam, validationOptionsMsg('Sort must be an enum'))
+    @IsOptional()
+      sort?: SortQAUParam;
+
+    @ApiPropertyOptional({
+      description: 'State of the user',
+      enum: State,
```
Comment (f-ogoog) [inline src/v2/api/dtos/QueryAllUsersDTO.ts:30]: add type

File: src/v2/api/dtos/QueryAllUsersDTO.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    })
+    @Transform(({ value }) => value.split(','))
+    @IsOptional()
+      state?: State[];
```
Comment (f-ogoog) [inline src/v2/api/dtos/QueryAllUsersDTO.ts:36]: now you don't validate parameter, you should add @Validate or add validation in Pipe

File: src/v2/api/dtos/QueryAllUsersDTO.ts [hunk 7]
```diff
@@ -0,0 +1,28 @@
...
+    })
+    @IsEnum(SortQAUParam, validationOptionsMsg('Sort must be an enum'))
+    @IsOptional()
+      sort?: SortQAUParam;
+
+    @ApiPropertyOptional({
+      description: 'State of the user',
+      enum: State,
+    })
+    @Transform(({ value }) => value.split(','))
```
Comment (f-ogoog) [inline src/v2/api/dtos/QueryAllUsersDTO.ts]: you can receive array in query so you don't need receive string

File: src/v2/api/dtos/QueryAllUsersDTO.ts [hunk 8]
```diff
@@ -0,0 +1,28 @@
+import { IsEnum, IsOptional } from 'class-validator';
+import { ApiPropertyOptional } from '@nestjs/swagger';
+import { State } from '@prisma/client';
+import { QueryAllDTO } from '../../utils/QueryAllDTO';
+import { validationOptionsMsg } from '../../utils/GLOBALS';
+import { SortQARParam } from './SortQARParam';
+import { SortQAUParam } from './SortQAUParam';
+import { Transform } from 'class-transformer';
+
+export class QueryAllUsersDTO extends QueryAllDTO {
```
Comment (f-ogoog) [inline src/v2/api/dtos/QueryAllUsersDTO.ts:12]: add search and write in description by which fields you serching

File: src/v2/api/responses/UsersResponse.ts [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import { ShortUserResponse } from './UserResponse';
+import { ApiProperty } from '@nestjs/swagger';
+import { State } from '@prisma/client';
+import { PaginationDataResponse } from './PaginationDataResponse';
+
+export class UserForGetAllResponse extends ShortUserResponse {
+    @ApiProperty({
+      description: 'User\'s telegram username',
```
Comment (f-ogoog) [inline src/v2/api/responses/UsersResponse.ts]: why telegram? it is fa username

File: src/v2/api/responses/UsersResponse.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    })
+      state: State;
+}
+
+export class UsersResponse {
+    @ApiProperty({
+      description: 'An array of information about the user',
+      type: [UserForGetAllResponse],
+    })
+      data: UserForGetAllResponse[];
```
Comment (f-ogoog) [inline src/v2/api/responses/UsersResponse.ts:25]: should be users

File: src/v2/api/services/UserService.ts [hunk 11]
```diff
@@ -577,4 +580,25 @@ export class UserService {
     await this.checkExcessiveSelectiveDisciplines(sortedDisciplines, sortedSelectedDisciplines, groupId);
     await this.attachSelectiveDisciplines(userId, body.disciplines);
   }
+
+  async getAll (query: QueryAllUsersDTO) {
+    const search = DatabaseUtils.getSearch(query, 'username');
```
Comment (f-ogoog) [inline src/v2/api/services/UserService.ts]: you should search also by email

File: src/v2/api/services/UserService.ts [hunk 12]
```diff
@@ -577,4 +580,25 @@ export class UserService {
...
+    const sort = DatabaseUtils.getSort(query, 'username');
+
+    const data: Prisma.UserFindManyArgs = {
+      where: {
+        ...search,
+        state: query.state? {
```
Comment (f-ogoog) [inline src/v2/api/services/UserService.ts]: query.state it's array so condition will be always true

File: src/v2/database/repositories/UserRepository.ts [hunk 13]
```diff
@@ -37,11 +37,12 @@ export class UserRepository {
     });
   }
 
-  async findMany (where: Prisma.UserWhereInput): Promise<DbUser[]> {
+  findMany (args?: Prisma.UserFindManyArgs): Promise<DbUser[]> {
+    console.log(args.where.state, 'args');                                         // ---------------LOG HERE!!!!!!!!!!!!!!-----------------
```
Comment (f-ogoog) [inline src/v2/database/repositories/UserRepository.ts]: mmm huita

File: src/v2/api/services/UserService.ts [hunk 14]
```diff
@@ -577,4 +580,25 @@ export class UserService {
...
+          in: query.state,
+        } : undefined,
+      },
+      ...sort,
+    };
+    const users = await DatabaseUtils.paginate<DbUser>(this.userRepository, query, data);
+
+    return {
```
Comment (f-ogoog) [inline src/v2/api/services/UserService.ts]: transfer it to the controller

File: src/v2/database/repositories/UserRepository.ts [hunk 15]
```diff
@@ -80,4 +81,10 @@ export class UserRepository {
       data,
     });
   }
+
+  async count (data: Prisma.UserCountArgs) {
+    return this.prisma.user.count(
+      data,
```
Comment (f-ogoog) [inline src/v2/database/repositories/UserRepository.ts]: write in 1 line

File: src/v2/mappers/UserMapper.ts [hunk 16]
```diff
@@ -13,4 +14,14 @@ export class UserMapper {
...
   }
+
+  getAll (users: DbUser[]) {
+    return users.map((user) => ({
+      id: user.id,
+      username: user.username,
+      email: user.email,
+      state: user.state,
+    })
+    );
```
Comment (f-ogoog) [inline src/v2/mappers/UserMapper.ts]: needless enter

File: src/v2/api/dtos/QueryAllUsersDTO.ts [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+      sort?: SortQAUParam;
+
+    @ApiPropertyOptional({
+      description: 'State of the user',
+      enum: State,
+    })
+    @Transform(({ value }) => value.split(','))
+    @IsOptional()
+      state?: State[];
+
```
Comment (f-ogoog) [inline src/v2/api/dtos/QueryAllUsersDTO.ts]: needless enter

**Review by f-ogoog — CHANGES_REQUESTED** [review]

File: src/v2/api/controllers/UserController.ts [hunk 18]
```diff
@@ -392,6 +385,36 @@ export class UserController {
...
+    type: UsersResponse,
+  })
+  @ApiBadRequestResponse({
+    description: `\n
+    InvalidQueryException:
+      Page must be a number
+      PageSize must be a number
+      Wrong value for order
+      Sort must be an enum
+      State must be an enum
```
Comment (f-ogoog) [inline src/v2/api/controllers/UserController.ts]: this message doesn't exist

**Review by f-ogoog — APPROVED** [review]

File: src/v2/api/services/UserService.ts [hunk 19]
```diff
@@ -577,4 +580,20 @@ export class UserService {
...
   }
+
+  async getAll (query: QueryAllUsersDTO) {
+    const search = DatabaseUtils.getSearch(query, 'username', 'email');
+    const sort = DatabaseUtils.getSort(query, 'username');
+
+    const data: Prisma.UserFindManyArgs = {
+      where: {
+        ...search,
+        state: query.state.length !== 0 ? {
```
Comment (fujivara) [inline src/v2/api/services/UserService.ts]: Since state is optional, check if there is a length property on it.

**Review by fujivara — CHANGES_REQUESTED** [review]

**Review by fujivara — APPROVED** [review]

**Comment by fokaaas** [comment]
closed as outdated

**Pull request closed without merging**